### PR TITLE
C5 gateway implementation merge

### PR DESCRIPTION
**Pull request by tharindu1st**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/main/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/ThrottleStreamProcessor.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+import org.wso2.siddhi.query.api.expression.Expression;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *
+ */
```
Comment (bhathiya) [inline analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/main/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/ThrottleStreamProcessor.java]: empty java doc.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGateway.java [hunk 2]
```diff
@@ -118,4 +130,29 @@
...
+     * @param application {@link Application}
+     * @throws GatewayException If there is a failure in notifying creation to gateway
+     */
+    void addApplication(Application application) throws GatewayException;
+
+    /**
+     * Publish the update application change to Gateway
+     *
+     * @param application {@link Application}
+     * @throws GatewayException
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGateway.java]: Add method comments description

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/ThrottlePolicyDeploymentManager.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
+package org.wso2.carbon.apimgt.core.api;
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/ThrottlePolicyDeploymentManager.java]: Should start the file with the license header.

File: gateway-modules/apim-ballerina-native-component/src/test/resources/log4j.properties [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#
+# Copyright 2012 WSO2, Inc. (http://wso2.com)
```
Comment (prasa7) [inline gateway-modules/apim-ballerina-native-component/src/test/resources/log4j.properties]: Wrong license year

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/internal/BundleActivator.java [hunk 5]
```diff
@@ -69,6 +71,14 @@ protected void start(BundleContext bundleContext) {
         } catch (NamingException e) {
             log.error("Error occurred while jndi lookup", e);
         }
+
+        // deploying default policies
+        try {
+            ThrottlerUtil.addDefaultAdvancedThrottlePolicies();
+            log.debug("Checked default throttle policies successfully");
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/internal/BundleActivator.java]: Check whether the debug log is enabled.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java [hunk 6]
```diff
@@ -55,12 +57,7 @@
 
     public APIGatewayPublisherImpl() {
         config = ServiceReferenceHolder.getInstance().getAPIMConfiguration();
-
-        //TODO:Remove this once broker is integrated
         gwHome = System.getProperty("gwHome");
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java]: introduce a constant please.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/ThrottlerUtil.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+                defaultQuotaPolicy.setType(PolicyConstants.REQUEST_COUNT_TYPE);
+                defaultQuotaPolicy.setLimit(requestCountLimit);
+                apiPolicy.setDefaultQuotaPolicy(defaultQuotaPolicy);
+                policyDAO.addPolicy(APIMgtConstants.ThrottlePolicyConstants.API_LEVEL, apiPolicy);
+            }
+        }
+    }
+
+    public static boolean isPolicyExist(String policyLevel, String policyName) throws APIManagementException {
+        PolicyDAO policyDAO = DAOFactory.getPolicyDAO();
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/ThrottlerUtil.java]: Missing java doc comment

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/configuration/models/PolicyDeployerConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.wso2.carbon.kernel.annotations.Element;
+
+/**
+ * Class to hold PolicyDeployer configurations
+ */
+public class PolicyDeployerConfiguration {
+
+    @Element(description = "Server URL")
+    private String serverURL = "https://localhost:9445/services/";
```
Comment (isurulucky) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/configuration/models/PolicyDeployerConfiguration.java]: @ruks is the port 9445 correct?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 9]
```diff
@@ -1370,6 +1370,50 @@ public void updateRating(String apiId, String ratingId, Rating rating) throws AP
...
+            try (ResultSet resultSet = statement.executeQuery()) {
+                if (resultSet.next()) {
+                    InputStream inputStream = resultSet.getBinaryStream(APIMgtConstants.GATEWAY_CONFIG);
+                    if (inputStream != null) {
+                        return IOUtils.toString(inputStream, StandardCharsets.UTF_8);
+                    } else {
+                        return "";
+                    }
+                } else {
+                    return "";
```
Comment (isurulucky) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: Need to throw and handle properly without returning an empty String.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIConfigContext.java [hunk 10]
```diff
@@ -20,60 +20,56 @@
...
-        this.createdTime = createdTime;
+
+    public APIConfigContext(API api, String packageName) {
+        this.api = api;
         this.packageName = packageName;
     }
 
+    public APIConfigContext(CompositeAPI compositeAPI, String gatewayPackageName) {
+
+        this.api = new API.APIBuilder(compositeAPI.getProvider(), compositeAPI.getName(), compositeAPI.getVersion())
```
Comment (uvindra) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/APIConfigContext.java]: This is a hack! Copying the values of a Composite API into a Standard API is misleading. 

It is better to only store the fields that are needed by the gateway without having a direct dependency on the entire API model object. This allows us to store the common information in the Composite API and Standard API model objects in the same fields without storing other redundant information. Please dont use hacks for new code!

File: analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/pom.xml [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.carbon.apimgt.throttling.siddhi.extension.*,
+                        </Export-Package>
+                        <Import-Package>
+                            org.apache.log4j.*,
```
Comment (harsha89) [inline analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/pom.xml]: Can you put the import version ranges?

Comment (harsha89) [inline analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/pom.xml]: This is for all imports

File: analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/main/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/ThrottleStreamProcessor.java [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+                                + attributeExpressionExecutors[2].getReturnType());
+            }
+        } else {
+            throw new ExecutionPlanValidationException("Throttle batch window should only have one/two parameter "
+                    + "(<int|long|time> windowTime (and <int|long> startTime), but found "
+                    + attributeExpressionExecutors.length + " input attributes");
+        }
+
+        List<Attribute> attributeList = new ArrayList<Attribute>();
+        attributeList.add(new Attribute("expiryTimeStamp", Attribute.Type.LONG));
```
Comment (harsha89) [inline analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/main/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/ThrottleStreamProcessor.java]: expiryTimeStamp -> to a constant

File: analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/main/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/ThrottleStreamProcessor.java [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+                long currentTime = executionPlanContext.getTimestampGenerator().currentTime();
+                if (startTime != -1) {
+                    expireEventTime = addTimeShift(currentTime);
+                } else {
+                    expireEventTime = executionPlanContext.getTimestampGenerator().currentTime() + timeInMilliSeconds;
+                }
+                if (scheduler != null) {
+                    scheduler.notifyAt(expireEventTime);
+                } else {
+                    log.error("scheduler is not initiated");
```
Comment (harsha89) [inline analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/main/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/ThrottleStreamProcessor.java]: should we thriw  a exception

File: analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/main/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/ThrottleStreamProcessor.java [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+                }
+            }
+            long currentTime = executionPlanContext.getTimestampGenerator().currentTime();
+            boolean sendEvents;
+            if (currentTime >= expireEventTime) {
+                expireEventTime += timeInMilliSeconds;
+                if (scheduler != null) {
+                    scheduler.notifyAt(expireEventTime);
+                } else {
+                    log.error("scheduler is not initiated");
```
Comment (harsha89) [inline analyzer-modules/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/main/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/ThrottleStreamProcessor.java]: should we throw a exception?

File: components/apimgt/org.wso2.carbon.apimgt.core/pom.xml [hunk 15]
```diff
@@ -909,7 +909,7 @@
             org.apache.commons.io.*;
             org.apache.commons.*;
             org.apache.velocity.*;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/pom.xml]: import version ranges?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIMgtAdminServiceImpl.java [hunk 16]
```diff
@@ -259,4 +249,46 @@ public String getAPIGatewayServiceConfig(String apiId) throws APIConfigRetrieval
...
 
+    @Override
+    public RegistrationSummary getRegistrationSummary() {
+        return new RegistrationSummary(apimConfiguration);
+    }
+
+    @Override
+    public List<Application> getAllApplications() throws APIManagementException {
+        try {
+            List<Application> applicationList = applicationDAO.getAllApplications();
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIMgtAdminServiceImpl.java]: we dont need to declare a varaible unless we do something

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIMgtAdminServiceImpl.java [hunk 17]
```diff
@@ -259,4 +249,46 @@ public String getAPIGatewayServiceConfig(String apiId) throws APIConfigRetrieval
...
+            String msg = "Error occurred while getting the Application list";
+            log.error(msg, ex);
+            throw new APIManagementException(msg, ExceptionCodes.APIM_DAO_EXCEPTION);
+        }
+    }
+
+    @Override
+    public List<String> getAllEndpoints() throws APIManagementException {
+        try {
+            List<String> endpointList = apiDAO.getUUIDsOfGlobalEndpoints();
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIMgtAdminServiceImpl.java]: we dont need to declare a varaible unless we do something

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIMgtAdminServiceImpl.java [hunk 18]
```diff
@@ -259,4 +249,46 @@ public String getAPIGatewayServiceConfig(String apiId) throws APIConfigRetrieval
...
+            String msg = "Error occurred while getting the Endpoint list";
+            log.error(msg, ex);
+            throw new APIManagementException(msg, ExceptionCodes.APIM_DAO_EXCEPTION);
+        }
+    }
+
+    @Override
+    public String getEndpointGatewayConfig(String endpointId) throws APIManagementException {
+        try {
+            String endpointConfig = apiDAO.getEndpointConfig(endpointId);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIMgtAdminServiceImpl.java]: we dont need to declare a varaible unless we do something

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 19]
```diff
@@ -218,12 +218,22 @@ public WorkflowResponse updateApplication(String uuid, Application application)
...
-                
+                        getWorkflowDAO(), getApiGateway());
                 application.setId(uuid);
                 application.setUpdatedUser(getUsername());
                 application.setUpdatedTime(LocalDateTime.now());
-                
+
+                String appTierName = application.getTier();
+                if (appTierName != null && !appTierName.equals(existingApplication.getTier())) {
+                    Policy policy = getPolicyDAO().getPolicy(APIMgtConstants.ThrottlePolicyConstants.APPLICATION_LEVEL,
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: when u break line mve remains from end of the =

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/configuration/models/PolicyDeployerConfiguration.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+import org.wso2.carbon.kernel.annotations.Element;
+
+/**
+ * Class to hold PolicyDeployer configurations
+ */
+public class PolicyDeployerConfiguration {
+
+    @Element(description = "Server URL")
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/configuration/models/PolicyDeployerConfiguration.java]: The description should mention which server url.

**Pull request merged**